src/keys/slip10.rs

```
@@ -399,88 +433,89 @@ impl<K: hazmat::Derivable> TryFrom<&[u8; 65]> for Slip10<K> {
-        Self(is.into_iter().map(Segment).collect())
-    }
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub struct Hardened(pub u32);
 
-    pub fn from_u32_hardened<I: IntoIterator<Item = u32>>(is: I) -> Self {
-        Self::from_u32(is.into_iter().map(|i| Segment::HARDEN_MASK | i))
+impl From<Hardened> for u32 {
+    fn from(segment: Hardened) -> u32 {
+        segment.0
     }
+}
 
-    pub fn join<O: AsRef<Chain>>(&self, o: O) -> Self {
-        let mut ss = self.0.clone();
-        ss.extend_from_slice(&o.as_ref().0);
-        Self(ss)
+impl TryFrom<u32> for Hardened {
+    type Error = ();
+    fn try_from(segment: u32) -> Result<Self, ()> {
+        if segment.is_hardened() {
+            Ok(Hardened(segment))
+        } else {
+            Err(())
+        }
     }
+}
 
-    pub fn is_empty(&self) -> bool {
-        self.0.is_empty()
+impl Segment for Hardened {
+    fn is_hardened(self) -> bool {
+        true
     }
-
-    pub fn len(&self) -> usize {
-        self.0.len()
+    fn harden(self) -> Hardened {
+        self
     }
-
-    pub fn segments(&self) -> &[Segment] {
-        &self.0
+    fn unharden(self) -> NonHardened {
+        NonHardened(self.0 ^ HARDEN_MASK)
     }
+}
 
-    pub fn all_hardened(&self) -> bool {
-        self.0.iter().all(Segment::is_hardened)
-    }
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub struct NonHardened(pub u32);
 
-    pub fn all_non_hardened(&self) -> bool {
-        !self.0.iter().any(Segment::is_hardened)
+impl From<NonHardened> for u32 {
+    fn from(segment: NonHardened) -> u32 {
+        segment.0
     }
 }
 
-impl Default for Chain {
-    fn default() -> Self {
-        Chain::empty()
+impl TryFrom<u32> for NonHardened {
+    type Error = ();
+    fn try_from(segment: u32) -> Result<Self, ()> {
+        if !segment.is_hardened() {
+            Ok(NonHardened(segment))
+        } else {
+            Err(())
```

Same here, should return an actual Error